### PR TITLE
vpc/vsw support DryRun check

### PR DESCRIPTION
**Pull request by linxudong7**
lxd@lxd-VirtualBox:~/gopath/src/github.com/terraform-providers/terraform-provider-alicloud$ TF_ACC=1 go test ./alicloud -v -run=TestAccAlicloudVpc -timeout=300m
=== RUN   TestAccAlicloudVpcsDataSourceBasic
--- PASS: TestAccAlicloudVpcsDataSourceBasic (64.35s)
=== RUN   TestAccAlicloudVpcBasic
--- PASS: TestAccAlicloudVpcBasic (28.82s)
=== RUN   TestAccAlicloudVpcMulti
--- PASS: TestAccAlicloudVpcMulti (42.00s)
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	135.195s
lxd@lxd-VirtualBox:~/gopath/src/github.com/terraform-providers/terraform-provider-alicloud$ TF_ACC=1 go test ./alicloud -v -run=TestAccAlicloudVSwitch -timeout=300m
=== RUN   TestAccAlicloudVSwitchesDataSourceBasic
--- PASS: TestAccAlicloudVSwitchesDataSourceBasic (61.26s)
=== RUN   TestAccAlicloudVSwitchBasic
--- PASS: TestAccAlicloudVSwitchBasic (30.50s)
=== RUN   TestAccAlicloudVSwitchMulti
--- PASS: TestAccAlicloudVSwitchMulti (29.48s)
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	121.271s

### CODE REVIEW

File: alicloud/data_source_alicloud_vpcs.go [hunk 1]
```diff
@@ -119,13 +119,22 @@ func dataSourceAlicloudVpcsRead(d *schema.ResourceData, meta interface{}) error
...
 	request.PageNumber = requests.NewInteger(1)
 
+	_, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+		request.DryRun = requests.NewBoolean(true)
+		return vpcClient.DescribeVpcs(request)
+	})
+	if err != nil && IsExceptedError(err, ForbiddenRAM) {
+		return WrapErrorf(err, DataDefaultErrorMsg, "alicloud_vpcs", request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+
```
Comment (ddaddaddda) [inline alicloud/data_source_alicloud_vpcs.go]: addDebug

File: alicloud/data_source_alicloud_vswitches.go [hunk 2]
```diff
@@ -136,12 +136,21 @@ func dataSourceAlicloudVSwitchesRead(d *schema.ResourceData, meta interface{}) e
...
 	}
 
+	_, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+		request.DryRun = requests.NewBoolean(true)
+		return vpcClient.DescribeVSwitches(request)
+	})
+	if err != nil && IsExceptedError(err, ForbiddenRAM) {
+		return WrapErrorf(err, DataDefaultErrorMsg, "alicloud_vswitches", request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+
```
Comment (ddaddaddda) [inline alicloud/data_source_alicloud_vswitches.go]: addDebug

File: alicloud/data_source_alicloud_vpcs.go [hunk 3]
```diff
@@ -119,13 +119,22 @@ func dataSourceAlicloudVpcsRead(d *schema.ResourceData, meta interface{}) error
 	request.PageSize = requests.NewInteger(PageSizeLarge)
 	request.PageNumber = requests.NewInteger(1)
 
+	_, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
```
Comment (ddaddaddda) [inline alicloud/data_source_alicloud_vpcs.go]: 为啥要提前做一次请求

File: alicloud/service_alicloud_vpc.go [hunk 4]
```diff
@@ -81,9 +81,20 @@ func (s *VpcService) DescribeVpc(id string) (v vpc.DescribeVpcAttributeResponse,
...
+	raw, err := s.client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+		request.DryRun = requests.NewBoolean(true)
+		return vpcClient.DescribeVpcAttribute(request)
+	})
+	if err != nil && IsExceptedError(err, ForbiddenRAM) {
+		response, _ := raw.(*vpc.DescribeVpcAttributeResponse)
+		v = *response
+		return v, WrapErrorf(err, DefaultErrorMsg, id, request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+
```
Comment (ddaddaddda) [inline alicloud/service_alicloud_vpc.go]: debug

File: alicloud/service_alicloud_vpc.go [hunk 5]
```diff
@@ -107,9 +118,20 @@ func (s *VpcService) DescribeVSwitch(id string) (v vpc.DescribeVSwitchAttributes
 	request := vpc.CreateDescribeVSwitchAttributesRequest()
 	request.VSwitchId = id
 
+	raw, err := s.client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+		request.DryRun = requests.NewBoolean(true)
+		return vpcClient.DescribeVSwitchAttributes(request)
+	})
+	if err != nil && IsExceptedError(err, ForbiddenRAM) {
+		response, _ := raw.(*vpc.DescribeVSwitchAttributesResponse)
+		v = *response
```
Comment (ddaddaddda) [inline alicloud/service_alicloud_vpc.go]: 出现错误转换response干嘛？

File: alicloud/data_source_alicloud_vpcs.go [hunk 6]
```diff
@@ -119,13 +119,23 @@ func dataSourceAlicloudVpcsRead(d *schema.ResourceData, meta interface{}) error
 	request.PageSize = requests.NewInteger(PageSizeLarge)
 	request.PageNumber = requests.NewInteger(1)
 
+	raw, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+		request.DryRun = requests.NewBoolean(true)
```
Comment (xiaozhu36) [inline alicloud/data_source_alicloud_vpcs.go]: I think the `Describe` does not need DryRun.

File: alicloud/data_source_alicloud_vswitches.go [hunk 7]
```diff
@@ -136,12 +136,22 @@ func dataSourceAlicloudVSwitchesRead(d *schema.ResourceData, meta interface{}) e
 		}
 	}
 
+	raw, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+		request.DryRun = requests.NewBoolean(true)
```
Comment (xiaozhu36) [inline alicloud/data_source_alicloud_vswitches.go]: still applies

File: alicloud/service_alicloud_vpc.go [hunk 8]
```diff
@@ -81,9 +81,19 @@ func (s *VpcService) DescribeVpc(id string) (v vpc.DescribeVpcAttributeResponse,
 	request := vpc.CreateDescribeVpcAttributeRequest()
 	request.VpcId = id
 
+	raw, err := s.client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+		request.DryRun = requests.NewBoolean(true)
+		return vpcClient.DescribeVpcAttribute(request)
+	})
+	if err != nil && IsExceptedError(err, ForbiddenRAM) {
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_vpc.go]: There is no need to catch `ForbiddenRAM`

**Pull request closed without merging**